### PR TITLE
add support different layer order in conformer

### PR DESCRIPTION
**Pull request by yqwangustc**
Initial experiments and past experiments (such as https://arxiv.org/pdf/2011.10798) show the `lconv_before_mhsa` can be helpful.

### CODE REVIEW

**Comment by yqwangustc** [comment]
Internal PR #11617 shows green.

File: axlearn/common/conformer.py [hunk 1]
```diff
@@ -202,6 +202,11 @@ class Config(BaseLayer.Config):
         )
         lconv: LConvLayer.Config = LConvLayer.default_config()
         norm: LayerNorm.Config = LayerNorm.default_config()
+        # Layer order. If None, default to "mhsa_before_conv", i.e., conformer layer order as
+        # secified in https://arxiv.org/abs/2005.08100.
+        # If not None, only "lconv_before_ff" "lconv_before_mhsa" "mhsa_before_lconv" is allowed.
```
Comment (markblee) [inline axlearn/common/conformer.py:207]:
```suggestion
        # If not None, only "lconv_before_ff" "lconv_before_mhsa" "mhsa_before_lconv" are allowed.
```

You can also consider a Literal[...] typing.

File: axlearn/common/conformer.py [hunk 2]
```diff
@@ -253,6 +258,12 @@ def __init__(self, cfg: Config, *, parent: Module):
                 f"cfg.right_context must be greater or equal to 0, get {cfg.right_context}."
             )
 
+        if cfg.layer_order is not None:
+            supperted_layer_order = ["lconv_before_ff", "lconv_before_mhsa", "mhsa_before_lconv"]
+            if cfg.layer_order not in supperted_layer_order:
+                msg = f"Only {supperted_layer_order} is allowed, got {cfg.layer_order}"
+                raise ValueError(msg)
```
Comment (markblee) [inline axlearn/common/conformer.py:265]:
```suggestion
            supported_layer_order = ["lconv_before_ff", "lconv_before_mhsa", "mhsa_before_lconv"]
            if cfg.layer_order not in supported_layer_order:
                raise ValueError(f"Layer order must be one of {supported_layer_order}, got {cfg.layer_order}.")
```

**Review by markblee — APPROVED** [review]

File: axlearn/common/conformer.py (hunk 2 above)
Comment (yqwangustc) [reply axlearn/common/conformer.py:265]: Thanks. Done.

File: axlearn/common/conformer.py (hunk 1 above)
Comment (yqwangustc) [reply axlearn/common/conformer.py:207]: Thanks for the suggestion. Updated.

File: axlearn/common/conformer.py [hunk 3]
```diff
@@ -253,6 +259,11 @@ def __init__(self, cfg: Config, *, parent: Module):
                 f"cfg.right_context must be greater or equal to 0, get {cfg.right_context}."
             )
 
+        if cfg.layer_order is not None:
+            supperted_layer_order = ["lconv_before_ff", "lconv_before_mhsa", "mhsa_before_lconv"]
```
Comment (markblee) [inline axlearn/common/conformer.py:263]:
```suggestion
            supported_layer_order = ["lconv_before_ff", "lconv_before_mhsa", "mhsa_before_lconv"]
```
This doesn't look fixed?

File: axlearn/common/conformer.py [hunk 4]
```diff
@@ -253,6 +259,11 @@ def __init__(self, cfg: Config, *, parent: Module):
...
+            if cfg.layer_order not in supperted_layer_order:
+                raise ValueError("Only {supperted_layer_order} is allowed, got {cfg.layer_order}")
```
Comment (markblee) [inline axlearn/common/conformer.py:265]:
```suggestion
                raise ValueError(f"Only {supperted_layer_order} is allowed, got {cfg.layer_order}")
```
This should be a fstring

File: axlearn/common/conformer.py (hunk 3 above)
Comment (yqwangustc) [reply axlearn/common/conformer.py:263]: Sorry. Missing this one. Fixed now.

File: axlearn/common/conformer.py (hunk 4 above)
Comment (yqwangustc) [reply axlearn/common/conformer.py:265]: Thanks. Done.

File: Dockerfile [hunk 5]
```diff
@@ -90,6 +90,19 @@ ENV PIP_FIND_LINKS=https://storage.googleapis.com/jax-releases/libtpu_releases.h
 RUN pip install .[tpu]
 COPY . .
 
+################################################################################
```
Comment (markblee) [inline Dockerfile:93]: Can you rebase the files?

Comment (yqwangustc) [reply Dockerfile]: Done. Thanks!

**Review by markblee — APPROVED** [review]

**Pull request merged**